RFCs/0011-partiql-insert-update-spec.md

```diff
@@ -0,0 +1,761 @@
+        | <tuple value>
+      }
+   [ WHERE <condition> ]
+
+<do replace> ::= SET {
+        <attr values> [, <attr values>]...
+            | <tuple value>
+      }
+   [ WHERE <condition> ]
+
+<attr values> ::=  {
+            <attr name> = { <value expr> | DEFAULT }
+            | ( <attr name> [, <attr name>]... ) = <value>
+            | ( <attr name> [, <attr name>]... ) = ( <sub-select> )
+        }
+
+<primary key> ::= <attr name>
+
+<composite primary key> ::= <attr name>, <attr name> [, <attr name> ]...
+
+<value expr> ::= <partiql value>
+     | <sql value expr>
+
+<partiql value> ::= NULL
+    | <ion value>
+    | <tuple value>
+    | <collection value>
+
+<sql value expr> ::= <sql expression>
+
+<ion value> ::= <backtick> <ion literal> <backtick>
+
+<tuple value> ::= <l curly brace> <r curly brace>
+    | <l curly brace> <string value> : <partiql value> [, <string value> : <partiql value>]... <r curly brace>
+
+<collection value> ::= <array value>
+    | <bag value>
+
+<array value> ::= <left bracket> { <partiql value> [, <partiql value>]... } <right bracket>
+
+<bag value> ::= <opening bag> { <partiql value> [, <partiql value>]... } <closing bag>
+
+<sub-select> ::= <sfw query>
+
+(* See sfw_query in PartiQL spec. figure 3: BNF Grammar for PartiQL Queries:
+https://partiql.org/assets/PartiQL-Specification.pdf *)
+<sfw query>
+
+<index attr name> ::= <attr name>
+<attr name> ::= <identifier>
+<constraint name> ::= <identifier>
+
+(* Literals *)
+<l curly brace> ::=
+“{”
+<r curly brace> ::=
+“}”
+<left bracket> ::=
+“[”
+<right bracket> ::=
+“]”
+<backtick> ::=
+`
+<opening bag> ::=
+<<
+<closing bag> ::=
+>>
+<left paren> ::=
+  “(”
+ <right paren> ::=
+  “)”
+```
+
+## 3. Description
+
+Implementation based on this specification MUST lead to atomic insertions of PartiQL values into a Schema (E.g. a database table). Atomic insertions means that either all of the insertions specified by the INSERT statement are performed, or none of the insertions are performed—E.g. when there is an error in insertion of one or more data items.
+
+### 3. 1 Attribute names
+
+As providing attribute names is optional, when attribute names are omitted and values are provided by `VALUES` clause or a `<sub-select>`:
+
+
+* (a) If the number of values in a data item is equal to the number of declared attributes, the values are assumed to refer to the declared attributes, in order of their definition in the table schema, and
+
+* (b) If the number of values in data item (M) is fewer than the number of declared attributes (N) the values are assumed to refer to the first M declared attributes, in order, and
+
+* (c) If the number of values in a data item (M) is greater than the number of declared attributes (N) it is an error, and
+* (d) If a value is not specified for every required attribute, it is an error.
+
+
+It’s worth emphasizing that, the implementing database MUST provide an ordering for the declared attributes that appear in the Schema.
```

Comment:
There should be some discussion here (ish) about how the `INSERT` statement is restricted when working in a schemaless fashion.  

For example, in

```sql
INSERT INTO (id, name) VALUES (1, 'bob');
```

The list of fields preceding `VALUES` is optional according to the grammar.  But if you don't have schema, the list of fields is absolutely required.

Reply:
Can you clarify what do you mean by not having schema based on the terms we've defined here? Is is an open-schema with no _declared attributes_? b/c in that case it falls into category `(c)` which is an error:
> (c) If the number of values in a data item (M) is greater than the number of declared attributes (N) it is an error, and